File: src/semp_workflow/modules/acl_profile.py
"""ACL Profile module - acl_profile.add, acl_profile.delete."""

from __future__ import annotations

import logging

from ..exceptions import SEMPError
from ..models import ActionResult, ResultStatus
from ..semp.client import SempClient
from ..semp.helpers import check_name_length, clean_payload, enc
from .base import BaseModule

logger = logging.getLogger(__name__)


class AclProfileAdd(BaseModule):
    description = "Create an ACL profile on the message VPN. Skipped if the profile already exists."
    params = {
        "aclProfileName":              {"type": "string", "required": True,  "description": "Name of the ACL profile"},
        "clientConnectDefaultAction":  {"type": "string", "required": False, "description": "Default action for client connections", "default": "disallow", "enum": ["allow", "disallow"]},
        "publishTopicDefaultAction":   {"type": "string", "required": False, "description": "Default action for publish topic exceptions", "default": "disallow", "enum": ["allow", "disallow"]},
        "subscribeTopicDefaultAction": {"type": "string", "required": False, "description": "Default action for subscribe topic exceptions", "default": "disallow", "enum": ["allow", "disallow"]},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        name = args.get("aclProfileName", "")
        if not name:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: aclProfileName")

        if err := check_name_length("aclProfileName", name):
            return ActionResult(ResultStatus.FAILED, err)

        path = f"aclProfiles/{enc(name)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking ACL profile: {e}")

        if exists:
            return ActionResult(ResultStatus.SKIPPED, f"ACL profile '{name}' already exists")

        if dry_run:
            return ActionResult(ResultStatus.DRYRUN, f"Would create ACL profile '{name}'")

        try:
            payload = clean_payload(args)
            client.create("aclProfiles", payload)
            return ActionResult(
                ResultStatus.OK, f"ACL profile '{name}' created"            )
        except SEMPError as e:
            return ActionResult(
                ResultStatus.FAILED, f"Failed to create ACL profile '{name}': {e}"
            )


class AclProfileDelete(BaseModule):
    description = "Delete an ACL profile from the message VPN. Skipped if the profile does not exist."
    params = {
        "aclProfileName": {"type": "string", "required": True, "description": "Name of the ACL profile to delete"},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        name = args.get("aclProfileName", "")
        if not name:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: aclProfileName")

        if err := check_name_length("aclProfileName", name):
            return ActionResult(ResultStatus.FAILED, err)

        path = f"aclProfiles/{enc(name)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking ACL profile: {e}")

        if not exists:
            return ActionResult(ResultStatus.SKIPPED, f"ACL profile '{name}' does not exist")

        if dry_run:
            return ActionResult(ResultStatus.DRYRUN, f"Would delete ACL profile '{name}'")

        try:
            client.delete(path)
            return ActionResult(
                ResultStatus.OK, f"ACL profile '{name}' deleted"            )
        except SEMPError as e:
            return ActionResult(
                ResultStatus.FAILED, f"Failed to delete ACL profile '{name}': {e}"
            )


MODULES = {
    "acl_profile.add": AclProfileAdd,
    "acl_profile.delete": AclProfileDelete,
}

File: internal/modules/acl_profile.go
package modules

import (
	"fmt"

	"semp-workflow/internal/models"
	"semp-workflow/internal/semp"
)

func init() {
	register("acl_profile.add", aclProfileAdd{})
	register("acl_profile.delete", aclProfileDelete{})
}

type aclProfileAdd struct{}

func (aclProfileAdd) Description() string {
	return "Create an ACL profile on the message VPN. Skipped if the profile already exists."
}

func (aclProfileAdd) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "aclProfileName", Type: "string", Required: true, Description: "Name of the ACL profile"},
		{Name: "clientConnectDefaultAction", Type: "string", Description: "Default action for client connections", Default: "disallow", Enum: []string{"allow", "disallow"}},
		{Name: "publishTopicDefaultAction", Type: "string", Description: "Default action for publish topic exceptions", Default: "disallow", Enum: []string{"allow", "disallow"}},
		{Name: "subscribeTopicDefaultAction", Type: "string", Description: "Default action for subscribe topic exceptions", Default: "disallow", Enum: []string{"allow", "disallow"}},
	}
}

func (aclProfileAdd) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	name := argStr(args, "aclProfileName", "")
	if name == "" {
		return failed("Missing required arg: aclProfileName")
	}
	if err := semp.CheckNameLength("aclProfileName", name); err != "" {
		return failed(err)
	}

	path := "aclProfiles/" + semp.Enc(name)

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking ACL profile: %s", err))
	}
	if exists {
		return skipped(fmt.Sprintf("ACL profile '%s' already exists", name))
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would create ACL profile '%s'", name))
	}

	if _, err := c.Create("aclProfiles", semp.CleanPayload(args)); err != nil {
		return failed(fmt.Sprintf("Failed to create ACL profile '%s': %s", name, err))
	}
	return ok(fmt.Sprintf("ACL profile '%s' created", name))
}

type aclProfileDelete struct{}

func (aclProfileDelete) Description() string {
	return "Delete an ACL profile from the message VPN. Skipped if the profile does not exist."
}

func (aclProfileDelete) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "aclProfileName", Type: "string", Required: true, Description: "Name of the ACL profile to delete"},
	}
}

func (aclProfileDelete) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	name := argStr(args, "aclProfileName", "")
	if name == "" {
		return failed("Missing required arg: aclProfileName")
	}
	if err := semp.CheckNameLength("aclProfileName", name); err != "" {
		return failed(err)
	}

	path := "aclProfiles/" + semp.Enc(name)

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking ACL profile: %s", err))
	}
	if !exists {
		return skipped(fmt.Sprintf("ACL profile '%s' does not exist", name))
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would delete ACL profile '%s'", name))
	}

	if err := c.Delete(path); err != nil {
		return failed(fmt.Sprintf("Failed to delete ACL profile '%s': %s", name, err))
	}
	return ok(fmt.Sprintf("ACL profile '%s' deleted", name))
}
